### PR TITLE
fix(light): make `OnDemand` generic instead of using the concrete type

### PR DESCRIPTION
**Pull request by niklasad1**
Helps to mock `OnDemand` to facilitate for example `RPC integration tests`

### CODE REVIEW

File: ethcore/light/src/on_demand/mod.rs [hunk 1]
```diff
@@ -94,6 +94,25 @@ pub mod error {
...
 
+/// Public interface for performing network requests `OnDemand`
+pub trait OnDemandRequester: Send + Sync {
+	/// Submit a strongly-typed batch of requests.
+	///
+	/// Fails if back-reference are not coherent.
+	fn request<T>(&self, ctx: &BasicContext, requests: T) -> Result<OnResponses<T>, basic_request::NoSuchOutput>
+	where
+		T: request::RequestAdapter,
+		Self: Sized;
```
Comment (ordian) [inline ethcore/light/src/on_demand/mod.rs]: I wonder why `Self: Sized` is required?

Comment (ascjones) [reply ethcore/light/src/on_demand/mod.rs]: Good question. Prevents being used as a trait object?

Comment (ordian) [reply ethcore/light/src/on_demand/mod.rs]: But this method is generic, I thought it's not allowed in a trait object already.

Comment (niklasad1) [reply ethcore/light/src/on_demand/mod.rs]: Thanks guys, it was not needed :+1: :1st_place_medal:

**Pull request merged**